server/stream.go

```diff
@@ -3115,45 +3127,57 @@ func (mset *stream) processJetStreamMsg(subject, reply string, hdr, msg []byte,
 
 // Internal message for use by jetstream subsystem.
 type jsPubMsg struct {
-	subj  string
-	dsubj string
+	dsubj string // Subject to send to, e.g. _INBOX.xxx
 	reply string
-	hdr   []byte
-	msg   []byte
-	o     *consumer
-	seq   uint64
+	StoreMsg
+	o *consumer
 }
 
 var jsPubMsgPool sync.Pool
 
 func newJSPubMsg(subj, dsubj, reply string, hdr, msg []byte, o *consumer, seq uint64) *jsPubMsg {
 	var m *jsPubMsg
+	var buf []byte
 	pm := jsPubMsgPool.Get()
 	if pm != nil {
 		m = pm.(*jsPubMsg)
+		buf = m.buf[:0]
 	} else {
-		m = &jsPubMsg{}
+		m = new(jsPubMsg)
 	}
 	// When getting something from a pool it is criticical that all fields are
 	// initialized. Doing this way guarantees that if someone adds a field to
 	// the structure, the compiler will fail the build if this line is not updated.
-	(*m) = jsPubMsg{subj, dsubj, reply, hdr, msg, o, seq}
+	(*m) = jsPubMsg{subj, reply, StoreMsg{dsubj, hdr, msg, buf, seq, 0}, o}
+
 	return m
 }
 
+// Gets a jsPubMsg from the pool.
+func getJSPubMsgFromPool() *jsPubMsg {
+	pm := jsPubMsgPool.Get()
+	if pm != nil {
+		return pm.(*jsPubMsg)
+	}
+	return new(jsPubMsg)
+}
+
 func (pm *jsPubMsg) returnToPool() {
 	if pm == nil {
 		return
 	}
 	pm.subj, pm.dsubj, pm.reply, pm.hdr, pm.msg, pm.o = _EMPTY_, _EMPTY_, _EMPTY_, nil, nil, nil
+	if len(pm.buf) > 0 {
+		pm.buf = pm.buf[:0]
+	}
 	jsPubMsgPool.Put(pm)
 }
 
 func (pm *jsPubMsg) size() int {
 	if pm == nil {
 		return 0
 	}
-	return len(pm.subj) + len(pm.reply) + len(pm.hdr) + len(pm.msg)
+	return len(pm.dsubj) + len(pm.reply) + len(pm.hdr) + len(pm.msg)
```

Comment:
This seem like a huge change, meaning that we could have gotten the size completely wrong. What was the impact of that?

Reply:
dsubj = "_INBOX.XX", subj = "foo", so could have been lower in our tests, but maybe higher in real use cases depending on length origin sub.

Reply:
Since you changed, does this need to be reverted?

Reply:
Ok, so I checked and we always call newJSPubMsg() with a dsubj (the now first param) and empty subj (the second). So not sure why we have the second in newJSPubMsg(), unless we think we may need it later.

Reply:
We used it for consumer msgs, but now they are constructed in a different way so could change that func.